platform/iphone/gamepad_iphone.mm

```diff
@@ -0,0 +1,276 @@
+#import "gamepad_iphone.h"
+#import "os_iphone.h"
+#import <GameController/GameController.h>
+
+@interface GamepadManager ()
+
+@property(nonatomic, strong) NSMutableDictionary *gamepadById;
+@property(nonatomic, strong) NSMutableArray *gamepads;
+@property(nonatomic, strong) NSMutableArray *pendingGamepads;
+
+@end
+
+@implementation GamepadManager
+
++ (instancetype)sharedManager {
+	static dispatch_once_t onceToken;
+	static GamepadManager *gGamepadManager = nil;
+	dispatch_once(&onceToken, ^{
+		gGamepadManager = [[GamepadManager alloc] init];
+	});
+	return gGamepadManager;
+}
+
+- (instancetype)init {
+	self = [super init];
+	if (self) {
+		self.gamepadById = [NSMutableDictionary dictionary];
+		self.gamepads = [NSMutableArray arrayWithCapacity:5];
+		self.pendingGamepads = [NSMutableArray arrayWithCapacity:3];
+
+		[[NSNotificationCenter defaultCenter]
+				addObserver:self
+				   selector:@selector(controllerWasConnected:)
+					   name:GCControllerDidConnectNotification
+					 object:nil];
+
+		[[NSNotificationCenter defaultCenter]
+				addObserver:self
+				   selector:@selector(controllerWasDisconnected:)
+					   name:GCControllerDidDisconnectNotification
+					 object:nil];
+	}
+	return self;
+}
+
+- (void)dealloc {
+	[[NSNotificationCenter defaultCenter]
+			removeObserver:self
+					  name:GCControllerDidConnectNotification
+					object:nil];
+
+	[[NSNotificationCenter defaultCenter]
+			removeObserver:self
+					  name:GCControllerDidDisconnectNotification
+					object:nil];
+
+	[super dealloc];
+}
+
+- (void)setReady:(BOOL)ready {
+	_ready = ready;
+
+	if (_ready) {
+		for (GCController *controller in self.pendingGamepads) {
+			[self addController:controller];
+		}
+		[self.pendingGamepads removeAllObjects];
+	}
+}
+
+- (GCControllerPlayerIndex)getFreePlayerIndex {
+	if (self.gamepads.count == 0) {
+		return GCControllerPlayerIndex1;
+	}
+	return (GCControllerPlayerIndex)self.gamepads.count - 1;
+}
+
+- (int)getJoyIdForController:(GCController *)controller {
+	NSNumber *key = [self.gamepadById allKeysForObject:controller].firstObject;
```

Comment:
This is kind of awkward, but allows us to map each controller back to the `joy_id` in Godot. This would make more sense if we flipped the keys and values in the dictionary. But unfortunately `GCController` does not conform to `NSCoding` and that is the only requirement to serve as a key in an `NSDictionary`